src/Microsoft.ML.Core/Data/VBufferMutationContext.cs

```diff
@@ -0,0 +1,148 @@
+                valuesCount,
+                keepOldOnResize: keepOldOnResize,
+                requireIndicesOnDense: requireIndicesOnDense);
+        }
+
+        internal static VBufferMutationContext<T> Create<T>(
+            ref VBuffer<T> destination,
+            int newLogicalLength,
+            int valuesCount,
+            int maxValuesCapacity)
+        {
+            return destination.GetMutableContext(
+                newLogicalLength,
+                valuesCount,
+                maxValuesCapacity);
+        }
+    }
+
+    /// <summary>
+    /// An object capable of mutation a <see cref="VBuffer{T}"/> by filling out
+    /// <see cref="Values"/> (and <see cref="Indices"/> if the buffer is not dense).
+    /// </summary>
+    public ref struct VBufferMutationContext<T>
+    {
+        private readonly int _logicalLength;
+        private readonly T[] _values;
+        private readonly int[] _indices;
+
+        /// <summary>
+        /// The mutable span of values.
+        /// </summary>
+        public readonly Span<T> Values;
+
+        /// <summary>
+        /// The mutable span of indices.
+        /// </summary>
+        public readonly Span<int> Indices;
+
+        /// <summary>
+        /// Gets a value indicating whether a new Values array was allocated.
+        /// </summary>
+        public bool CreatedNewValues { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether a new Indices array was allocated.
+        /// </summary>
+        public bool CreatedNewIndices { get; }
+
+        internal VBufferMutationContext(int logicalLength,
+            int physicalValuesCount,
+            T[] values,
+            int[] indices,
+            bool requireIndicesOnDense,
+            bool createdNewValues,
+            bool createdNewIndices)
+        {
+            _logicalLength = logicalLength;
+            _values = values;
+            _indices = indices;
+
+            bool isDense = logicalLength == physicalValuesCount;
+
+            Values = _values.AsSpan(0, physicalValuesCount);
+            Indices = !isDense || requireIndicesOnDense ? _indices.AsSpan(0, physicalValuesCount) : default;
+
+            CreatedNewValues = createdNewValues;
+            CreatedNewIndices = createdNewIndices;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="VBuffer{T}"/> using the current
+        /// Values and Indices.
+        /// </summary>
+        /// <param name="physicalValuesCount">
+        /// An optional size that allows reducing the number of physical values to be
+        /// represented in the created buffer.
+        /// This is useful in sparse situations where the mutation context was created
+        /// with a larger physical value count than was needed
+        /// because the final value count was not known at creation time.
+        /// </param>
+        /// <returns></returns>
+        public VBuffer<T> CreateBuffer(int? physicalValuesCount = null)
```

Comment:
>CreateBuffer [](start = 26, length = 12)

should we assert that this only called once? We don't want it to be called multiple times, right?

Also, if we don't, should we call it something like `Commit`? I can see you already had the name 'Complete' before, and you left it inside a comment.

Reply:
It's not possible to assert it is only called once because `VBufferMutationContext<T>` is a `ref struct` (and actually I need to mark it as `readonly`). Since it is a ref struct, it shouldn't change state because you can get in a lot of trouble passing it into methods as parameters, or making copies of it on the stack.

I can change the name of `CreateBuffer`, if you'd like.  Do you have a suggestion?  Do you like `Commit` more than `CreateBuffer`? I didn't like `Complete`, but I can be persuaded.

---
In reply to: [233652648](https://github.com/dotnet/machinelearning/pull/1580#discussion_r233652648) [](ancestors = 233652648)